drivers/rtc/rtc_pcf2123.c

```diff
@@ -0,0 +1,663 @@
+ *
+ * The PCF2123 starts from the address RA and auto-increments for every byte.
+ */
+#define PCF2123_CMD_READ  ((1U << 7) | (0b001 << 4))
+#define PCF2123_CMD_WRITE ((0U << 7) | (0b001 << 4))
+
+#ifdef PCF2123_INT1_GPIOS_IN_USE
+void callback_work_handler(struct k_work *work);
+K_WORK_DEFINE(callback_work, callback_work_handler);
+
+static void gpio_int1_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins);
+#endif
+
+struct pcf2123_config {
+	const struct spi_dt_spec spi;
+#ifdef PCF2123_INT1_GPIOS_IN_USE
+	struct gpio_dt_spec int1;
+#endif
+};
+
+struct pcf2123_data {
+#ifdef PCF2123_INT1_GPIOS_IN_USE
+	const struct device *dev;
+	rtc_alarm_callback alarm_callback;
+	void *alarm_user_data;
+	struct gpio_callback int1_callback;
+	struct k_work callback_work;
+#endif
+};
+
+/**
+ * @brief Read from a single or multiple PCF2123 registers.
+ *
+ * @param dev The RTC device.
+ * @param start_addr The first memory address to start reading from.
+ * @param buf Buffer to store read data.
+ * @param len Number of bytes to read.
+ * @retval 0 on success, else -ERRNO.
+ *
+ * @note The PCF2123 auto-increments the selected memory address for every byte sent.
+ */
+static int pcf2123_reg_read(const struct device *dev, const uint8_t start_addr, uint8_t *buf,
+			    const size_t len)
+{
+	if (!buf) {
+		LOG_ERR("Failed to read from start addr %d, buffer is NULL", start_addr);
+		return -EINVAL;
+	}
+	if (len <= 0) {
+		LOG_ERR("Failed to read from start addr %d, buffer length is %d", start_addr, len);
+		return -EINVAL;
+	}
+
+	const struct pcf2123_config *config = dev->config;
+
+	uint8_t cmd_byte = PCF2123_CMD_READ | start_addr;
+	struct spi_buf tx_buf = {.buf = &cmd_byte, .len = sizeof(cmd_byte)};
+	const struct spi_buf_set tx_bufs = {.buffers = &tx_buf, .count = 1};
+
+	uint8_t rx_dummy; /* Dummy byte. */
+	struct spi_buf rx_buf[2] = {{.buf = &rx_dummy, .len = 1}, {.buf = buf, .len = len}};
+	const struct spi_buf_set rx_bufs = {.buffers = rx_buf, .count = 2};
+
+	int ret = spi_transceive_dt(&config->spi, &tx_bufs, &rx_bufs);
+	if (ret < 0) {
+		LOG_ERR("Failed to read from register with start address %d (err %d)", start_addr,
+			ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Write to a single or multiple PCF2123 registers.
+ *
+ * @param dev The RTC device.
+ * @param start_addr The first memory address to start writing towards.
+ * @param buf Data to write.
+ * @param len Data size in bytes.
+ * @retval 0 on success, else -ERRNO.
+ *
+ * @note The PCF2123 auto-increments the selected memory address for every byte sent.
+ */
+static int pcf2123_reg_write(const struct device *dev, const uint8_t start_addr, uint8_t *buf,
+			     const size_t len)
+{
+	if (!buf) {
+		LOG_ERR("Failed to write to start addr %d, buffer is NULL", start_addr);
+		return -EINVAL;
+	}
+	if (len <= 0) {
+		LOG_ERR("Failed to write to start addr %d, buffer length is %d", start_addr, len);
+		return -EINVAL;
+	}
```

Comment:
These checks are redundant :) Its an internal function, so there is pretty much 0 chance an invalid buffer will be provided :) If you really want checks anyway, use __ASSERT_NO_MSG() since these are not compiled in for production code :)